pkg/kube/watcher.go

```diff
@@ -67,29 +70,56 @@ func (e *EventWatcher) OnUpdate(oldObj, newObj interface{}) {
 	e.onEvent(event)
 }
 
-// Ignore events older than the maxEventAgeSeconds
+// Ignore events older than the maxEventAgeSeconds.
 func (e *EventWatcher) isEventDiscarded(event *corev1.Event) bool {
-	timestamp := event.LastTimestamp.Time
-	if timestamp.IsZero() {
-		timestamp = event.EventTime.Time
-	}
-	eventAge := time.Since(timestamp)
+	eventAge, timeUsedForAge := getEventAge(event, clock.RealClock{})
 	if eventAge > e.maxEventAgeSeconds {
 		// Log discarded events if they were created after the watcher started
-		// (to suppres warnings from initial synchrnization)
-		if timestamp.After(startUpTime) {
+		// (to suppres warnings from initial synchronization)
+		if timeUsedForAge.After(startUpTime) {
 			log.Warn().
-				Str("event age", eventAge.String()).
-				Str("event namespace", event.Namespace).
-				Str("event name", event.Name).
-				Msg("Event discarded as being older then maxEventAgeSeconds")
+				Str("eventAge", eventAge.String()).
+				Str("eventNamespace", event.Namespace).
+				Str("eventName", event.Name).
+				Str("eventMessage", event.Message).
+				Str("eventReason", event.Reason).
+				Str("eventSourceComponent", event.Source.Component).
+				Str("eventSourceHost", event.Source.Host).
+				Str("eventFirstTimestamp", event.FirstTimestamp.String()).
+				Str("eventLastTimestamp", event.LastTimestamp.String()).
+				Str("eventCreationTime", event.CreationTimestamp.String()).
+				Str("eventEventTime", event.EventTime.String()).
+				Str("involvedObjectAPIVersion", event.InvolvedObject.APIVersion).
+				Str("involvedObjectKind", event.InvolvedObject.Kind).
+				Str("involvedObjectNamespace", event.InvolvedObject.Namespace).
+				Str("involvedObjectName", event.InvolvedObject.Name).
+				Str("involvedObjectUID", string(event.InvolvedObject.UID)).
+				Msg(fmt.Sprintf("Event discarded as being older than maxEventAgeSeconds: %v", e.maxEventAgeSeconds))
 			e.metricsStore.EventsDiscarded.Inc()
 		}
 		return true
 	}
 	return false
 }
 
+// getEventAge returns the age of the event and the time used for age calculation.
+// The time used for age calculation is the greater of creationTimestamp and
+// lastTimestamp. While based on event aggregation, lastTimestamp should be greater
+// than creationTimestamp, in practice it doesn't always seem to be the case.
+// Hence, we compare and use the greater of the two to ensure that we process the
+// event and don't skip it.
```

Comment:
I have a hypothesis. We see `creationTimestamp > lastTimestamp` because events get deleted from `etcd` while still in client caches. When the API is called, the event doesn't exist and a new event is created with the data with a brand new creation timestamp (as we know, `lastTimestamp` is generated by the client and `creationTimestamp` by the API server). If this hypothesis is right this change would not make a difference and we'll still see the discrepancy. But it also means there's might be no way clean to solve this besides keeping events for longer, which has other consequences.